### PR TITLE
Initial Gitlab CI support [gitlab-ci-dev]

### PR DESCRIPTION
**Pull request by adrienbernede**
This PR introduces a first implementation of a Gitlab pipeline for MFEM.

In particular, this pipeline is setup to:
- perform some static checks (code style and documentation)
- build MFEM prerequisites
- build MFEM for different configurations
- perform sanity checks
- run baseline test and share the diff

TODO:
- [X] use existing style test (introduced by #1244)
- [x] validate baseline update mechanism and restrict it to master branch
- CANCELLED: fix jobs triggered when pipeline failed early. [Bug in Gitlab](https://gitlab.com/gitlab-org/gitlab/issues/31526)
- [x] fix concurrency issue on baseline test (mfem link naming): will solve itself by adding baseline test in mfem (no need to call tests from an outside repo)
- [x] print diff in the logs (style check and baseline check)

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1203](https://github.com/mfem/mfem/pull/1203) | @adrienbernede | @tzanio | @acfisher + @tzanio | 12/19/19 | 03/18/20 | 03/19/20 | |

### CODE REVIEW

**Comment by adrienbernede** [comment]
LGTM

**Comment by tzanio** [comment]
@acfisher and @v-dobrev, reminder that the reviews for this PR are due in a week

**Comment by tzanio** [comment]
@mfem/reviewers -- this PR will affect the review process, so please take a look and provide your input to @adrienbernede and @acfisher.

**Comment by barker29** [comment]
I suggest updating `CONTRIBUTING.md` to give some basic info about how to use/check the Gitlab CI features.

**Comment by adrienbernede** [comment]
@barker29 
I updated `CONTRIBUTING.md`, please let me know if that’s what you expected.

**Comment by adrienbernede** [comment]
@tzanio @acfisher 

We are experiencing some issues with Gitlab mirroring.
Basically, since we filter out branches that are not "approved", Gitlab fails to create them and considers each failure as an error. Eventually, Gitlab blocks mirroring because of those recurring errors. This does not happen when the mirror repo has just been created. Which is why it appeared to be working in December.

In short, error handling in Gitlab is at fault here, and @tgmachina will be discussing possible improvements with the Gitlab team.

If this PR is accepted, I think we should still merge it. However the Gitlab CI would not be working until further notice.

In the meantime, we could create another MFEM repo on Gitlab (not a mirror repo), in which branches can be pushed to trigger the pipeline. This would be a manual process, but anyone with an LC Gitlab account could trigger tests when necessary. Scaling isn’t guaranteed though.

**Comment by tgmachina** [comment]
@adrienbernede 

Currently, the result of that error handling is that mirroring for the repo stops--would it suffice right now to re-enable it periodically in GitLab? It's also not ideal, but maybe easier for one person to just turn it back on in lieu of maintaining another repo that is having CI triggered on it manually, just a thought.

I spoke with the GitLab team today and they agree that this is a priority. I'll open an issue with them and we'll work on getting the fix made in GitLab server. Meanwhile, I could work on patching the offending code myself in our own instance.

**Comment by adrienbernede** [comment]
I first I thought that mirroring was paused if too many errors occurred during one mirroring attempt.
Umpire case proved me wrong. One error interrupts the mirroring. After several interruptions, the mirroring is paused.
Re-enabling mirroring would NOT make it to work again, even for a little while. I was also confused by the false positive we sometimes get (mirroring marked as successful in settings page, but marked as failing on project page).

File: makefile [hunk 1]
```diff
@@ -675,6 +678,16 @@ style:
 	   echo "No source files were changed.";\
 	fi
 
+style-check:
```
Comment (tzanio) [inline makefile]: I suggest we handle this by introducing `code-style` script in #1219 (see https://mybitbucket.llnl.gov/projects/MFEM/repos/tests/browse/code-style)

**Comment by tzanio** [comment]
Delaying by a week to give folks time to address the above issue.

File: makefile (hunk 1 above)
Comment (adrienbernede) [reply makefile]: @tzanio @acfisher 
It seems to me that the test you mention actually applies astyle change.
However in this PR, @acfisher introduced an "astyle check" that tells if style is OK.

I would rather use this check, because I don’t think it would be easy to modify the code during the PR testing. However it’s the perfect place to tell the developers they should have check their code for style compliance.

Is that OK?

**Comment by tzanio** [comment]
Delaying to give folks time to address issues.

**Comment by tzanio** [comment]
Merged in `master` to trigger new Travis checks...

**Comment by adrienbernede** [comment]
> Merged in `master` to trigger new Travis checks...

I was surprised to see "merged in master" in my emails...

**Comment by adrienbernede** [comment]
LGTM

**Comment by adrienbernede** [comment]
Let’s trigger tests on Gitlab too.

**Comment by adrienbernede** [comment]
Now using the same script for style check as in Travis.

@acfisher, @tzanio suggested to revert the (now unused) change in the makefile introducing style-check target (308861c6d09cd8c9d910f64d423a84f72b5c4f75).
Sounds good to me, can you please confirm that’s OK for you too?

**Comment by acfisher** [comment]
Yes, we now have another approach for that functionality.  Sounds good to me.

**Comment by adrienbernede** [comment]
Done!

**Comment by adrienbernede** [comment]
@acfisher @tzanio 
I’m not sure what we agreed on:
- Do you remove doc and style check from gitlab ci (sounds OK to me)
- Do we want to use the tests repo, or do we move baseline check to the new tests/scripts directory? If yes, maybe the tests repo can be converted to a baseline repo?

**Comment by tzanio** [comment]
LGTM

**Comment by tzanio** [comment]
This needs more testing and there still seem to be some issues... extending the review due date to 02/11/20.

**Comment by acfisher** [comment]
LGTM

**Comment by acfisher** [comment]
@adrienbernede it looks like the lib builds are failing?

**Comment by adrienbernede** [comment]
LGTM

**Comment by adrienbernede** [comment]
LGTM

**Comment by adrienbernede** [comment]
LGTM

File: tests/scripts/code-style [hunk 2]
```diff
@@ -48,8 +48,10 @@ shift
 
 # Verify C++ code styling in MFEM
 cd $mfem_dir
+
 make distclean
-make style
+make style ASTYLE_BIN=${ASTYLE:-astyle}
```
Comment (tzanio) [inline tests/scripts/code-style]: Why is this necessary?

File: makefile [hunk 3]
```diff
@@ -662,7 +662,8 @@ status info:
 	$(info MFEM_MPI_NP            = $(MFEM_MPI_NP))
 	@true
 
-ASTYLE = astyle --options=$(SRC)config/mfem.astylerc
+ASTYLE_BIN = astyle
+ASTYLE = $(ASTYLE_BIN) --options=$(SRC)config/mfem.astylerc
 FORMAT_FILES = $(foreach dir,$(DIRS) $(EM_DIRS) config,"$(dir)/*.?pp")
```
Comment (tzanio) [inline makefile]: This is OK in general, but why will the `astyle` binary have a different name?

File: .gitlab-ci.yml [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+    script:
+        - cd ${BUILD_ROOT}/tpls
+        - mkdir -p ${BUILD_PATH}
+        - |
+          if [[ "${MPIEXEC}" == "srun" ]]
+          then
+              export JOB_ID=$(squeue -h --name=${MFEM_ALLOC_NAME} --format=%A)
+              export EXEC_PREFIX="srun --jobid=${JOB_ID}"
+          fi
+        - $EXEC_PREFIX make -j 4 ${METIS} PREFIX=${BUILD_PATH}/ BUILD_DIR=${BUILD_PATH} LOG_DIR=${BUILD_PATH} CC=cc CXX=c++
```
Comment (tzanio) [inline .gitlab-ci.yml]: Shouldn't the `-j 4` and the `-j 8` below be machine-dependent?

File: .gitlab-ci.yml [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+    stage: test
+    variables:
+        GIT_STRATEGY: none
+    script:
+        - cd ${BUILD_PATH}/mfem_${MFEM_USE_MPI}_${MFEM_DEBUG}
+        - make test
+
+# The BUILD_ROOT is unique to the branch and the pipeline ID. Branch is not
+# necessary but makes it easier to retrieve a data. This means that there cannot
+# be concurrency between 2 pipelines.
```
Comment (tzanio) [inline .gitlab-ci.yml]: I don't understand what is meant by the second sentence above.

**Comment by tzanio** [comment]
Maybe that's too much to ask, but can we have a GitLab badge like the Travis and Appveyor one at https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md?

**Comment by tzanio** [comment]
LGTM

**Comment by tzanio** [comment]
> LGTM

@adrienbernede -- I don't see a run for this at https://lc.llnl.gov/gitlab/mfem/mfem/pipelines... am I missing something?

**Comment by tzanio** [comment]
Looking at this again, I wonder if it is better to merge as is with the `baseline` test only so the team can get used to running it, and add longer tests later?

**Comment by adrienbernede** [comment]
@tzanio, I should have shared my changes on Friday...
I changed the way I manage the allocation to allow the addition of sample-runs-baseline. The consequence is that it is much more difficult to get resource on daytime. But the run-samples-baseline are here, and I just have to wait hours between each "validation" cycle.

I also added some comments in the yaml files, but plan on describing the pipeline more thoroughly. What is the best place to document it?

If you want to merge the baseline only version, I can add a commit doing the necessary, removing the sample-runs-baseline and moving back to debug queue for more responsiveness.

**Comment by adrienbernede** [comment]
LGTM

**Comment by adrienbernede** [comment]
@tzanio, this is working for me, however I’m not sure why I have differences in baseline test (except because of timing).

**Comment by adrienbernede** [comment]
LGTM

File: .gitlab-ci.yml (hunk 4 above)
Comment (adrienbernede) [reply .gitlab-ci.yml]: Now configurable per machine

File: .gitlab-ci.yml (hunk 5 above)
Comment (adrienbernede) [reply .gitlab-ci.yml]: I am using 2 parameters to name the directory `mfem_${CI_COMMIT_REF_SLUG}_${CI_PIPELINE_ID}`:
- CI_COMMIT_REF_SLUG is the ref name (converted), usually the branch/tag name
- CI_PIPELINE_ID
Since the pipeline ID is unique, ref name is not necessary, but it make the directory name more human readable: I can easily tell where the working dir for my branch is.

File: makefile (hunk 3 above)
Comment (adrienbernede) [reply makefile]: When I was using astyle in Gitlab, I wanted to specify the full path to the desired version on astyle.
This is now useless and has been reverted.

File: tests/scripts/code-style (hunk 2 above)
Comment (adrienbernede) [reply tests/scripts/code-style]: When I was using astyle in Gitlab, I wanted to specify the full path to the desired version on astyle.
This is now useless and has been reverted.

**Comment by tzanio** [comment]
LGTM

**Comment by jandrej** [comment]
Merged in `next` for testing...

**Comment by tzanio** [comment]
Adding a few extra days for approval...

**Comment by adrienbernede** [comment]
@tzanio sorry for the forced push, I had a merge with master in my diverged history. I cherry-picked your changes. This time it was carefully done.

**Comment by adrienbernede** [comment]
LGTM

**Comment by adrienbernede** [comment]
Baseline and rebaseline have been extensively tested to check if it would work on master, and check if the commit works. I used a clone to emulate the behavior on master, and tested to rebaseline in another tests repository.

**Comment by tzanio** [comment]
Adding a bit more time to review...

**Comment by adrienbernede** [comment]
LGTM

**Comment by tzanio** [comment]
Is this ready to go @adrienbernede and @acfisher ?

**Comment by adrienbernede** [comment]
I am on a mirroring issue with this branch...

File: .gitlab-ci.yml [hunk 6]
```diff
@@ -286,7 +283,7 @@ stages:
           exit 1
       fi
     - git add baseline-${SYS_TYPE}.saved
-    - git commit -m "${SYS_TYPE} rebaselined in Gitlab pipeline ${CI_PIPELINE_ID}"
+    - git commit -m "${SYS_TYPE} rebaselined in Gitab pipeline ${CI_PIPELINE_ID}"
```
Comment (adrienbernede) [inline .gitlab-ci.yml]:
```suggestion
    - git commit -m "${SYS_TYPE} rebaselined in GitLab pipeline ${CI_PIPELINE_ID}"
```

**Comment by adrienbernede** [comment]
LGTM

File: .gitlab-ci.yml [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+  - libs
+  - build
+  - test
+  - deallocate
+  - baseline_check
+  - baseline_publish
+
+# The setup job in setup stage don’t rely on MFEM git repo. It prepares a
+# pipeline-wide working directory downloading/updating external repos.
+# TODO: update is not necessary anymore, repo are never shared with another
```
Comment (acfisher) [inline .gitlab-ci.yml]: Is this todo still required?  I don't see an update step here.

File: .gitlab/quartz.yml [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+    MFEM_DEBUG: "NO"
+    MFEM_USE_MPI: "YES"
+  needs: [build_mfem_opt_par_gcc_6_1_0_quartz, allocate_quartz]
+
+# Baseline
+baselinecheck_mfem_intel_quartz:
+  extends: [.baselinecheck_mfem, .on_quartz]
+  needs: [allocate_quartz]
+
+  #samplebaselinecheck_mfem_gcc_6_1_0_quartz:
```
Comment (acfisher) [inline .gitlab/quartz.yml]: Remove dead code.

**Comment by acfisher** [comment]
Is the mirroring up and running?  I don't see a pipeline launch under the latest magic word.

**Comment by adrienbernede** [comment]
> Is the mirroring up and running? I don't see a pipeline launch under the latest magic word.

It’s not working, I am talking with @tgmachina about this.

File: .gitlab-ci.yml (hunk 7 above)
Comment (adrienbernede) [reply .gitlab-ci.yml]: The comment is not clear enough:
It’s saying that in the setup job, it’s not necessary to update git repositories, or even to test if they already exist, because for each pipeline, we create another ${BUILD_ROOT}. I would like to leave such a comment to remember that this is not optimal, so we you should eventually decide to optimize this (once it’s robust) or assume the extra amount of data generate.
In general, I think we will need to address the fact we will be accumulating a lot of data.

File: .gitlab/quartz.yml (hunk 8 above)
Comment (adrienbernede) [reply .gitlab/quartz.yml]: This is indeed dead, but waiting to be resurrected :)

**Comment by tzanio** [comment]
LGTM

**Comment by adrienbernede** [comment]
@tzanio No change since the last mirroring, so sing-off won’t have effect here.
Let me push the changes requested by @acfisher so that you can play with it.

**Comment by tzanio** [comment]
LGTM

**Comment by adrienbernede** [comment]
The PR is now fully functional from my perspective, let’s see how it scales!

File: .gitlab/quartz.yml (hunk 8 above)
Comment (adrienbernede) [reply .gitlab/quartz.yml]: Done

File: .gitlab-ci.yml (hunk 7 above)
Comment (adrienbernede) [reply .gitlab-ci.yml]: Made the comment more clear.

**Comment by adrienbernede** [comment]
Reminder: Sign-off does not "trigger" CI, but allows it, so that the next time mirroring happens, CI will be triggered.
Mirroring can be forced on CI side, which I did to speed-up the process here.

**Comment by acfisher** [comment]
Looks like it is in good shape.  Lets get this pulled in asap.

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -370,6 +370,9 @@ Before you can start, you need a GitHub account, here are a few suggestions:
   checks in Travis enforce MFEM-specific rules which are explained in the
   error messages and the `tests/scripts` directory.
 
+- If triggered, track the status of the LLNL GitLab tests. If failing, ask
+  one of the _LLNL developer_ for details.
```
Comment (v-dobrev) [inline CONTRIBUTING.md]: typo: developer -> developers

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by v-dobrev** [comment]
I "parmetis" build in the pipeline -- not sure why it is there -- we can probably remove it.

**Comment by adrienbernede** [comment]
I think Tzanio mentioned it as well. I was assuming `MFEM_USE_MPI` implied `parmetis`. 
@acfisher Do you remember what was the intention?

**Comment by v-dobrev** [comment]
Some TPLs depend on parmetis but mfem does not depend on it directly. The parallel mfem build requires just serial metis.

**Comment by tzanio** [comment]
We can change that later... I think for now the goal is for folks to get experience with GitLab.
